design/general-progress-monitoring.md

```diff
@@ -0,0 +1,499 @@
+
+Post-snapshot and other operations may take a long time and Velero and its plugins may be restarted
+during this time.  Once a backup/restore has moved into the WaitingForPluginOperations or
+WaitingForPluginOperationsPartialFailure phase, another backup/restore may be started.
+
+While in the WaitingForPluginOperations or WaitingForPluginOperationsPartialFailure phase, the
+snapshots and item actions will be periodically polled.  When all of the snapshots and item actions
+have reported success, the backup/restore will move to the Completed or PartiallyFailed phase,
+depending on whether the backup/restore was in the WaitingForPluginOperations or
+WaitingForPluginOperationsPartialFailure phase.
+
+The Backup resources will not be written to object storage until the backup has entered a final phase: 
+Completed, Failed or PartialFailure
+
+## Reconciliation of InProgress backups
+
+InProgress backups will not have a `velero-backup.json` present in the object store.  During
+reconciliation, backups which do not have a `velero-backup.json` object in the object store will be
+ignored.
+
+## Plug-in API changes
+
+### OperationProgress struct
+
+    type OperationProgress struct {
+        completed bool                          // True when the operation has completed, either successfully or with a failure
+        err error                               // Set when the operation has failed
+        itemsCompleted, itemsToComplete int64   // The number of items that have been completed and the items to complete
+                                                // For a disk, an item would be a byte and itemsToComplete would be the
+                                                // total size to transfer (may be less than the size of a volume if
+                                                // performing an incremental) and itemsCompleted is the number of bytes
+                                                // transferred.  On successful completion, itemsCompleted and itemsToComplete
+                                                // should be the same
+        started, updated time.Time              // When the upload was started and when the last update was seen.  Not all
+                                                // systems retain when the upload was begun, return Time 0 (time.Unix(0, 0))
+                                                // if unknown.
+    }
+
+### VolumeSnapshotter changes
+
+Two new methods will be added to the VolumeSnapshotter interface:
+
+    Progress(snapshotID string) (UploadProgress, error)
+    Cancel(snapshotID string) (UploadProgress, error)
+
+Open question: Does VolumeSnapshotter need Cancel, or is that only needed for BIA/RIA?
+
+Progress will report the current status of a snapshot upload.  This should be callable at
+any time after the snapshot has been taken.  In the event a plug-in is restarted, if the operationID
+(snapshot ID) continues to be valid it should be possible to retrieve the progress.
```

Comment:
This is up to the plugin to work out. In the case of a plugin which creates a CR to signal an operation from an external controller (datamover use case), then the plugin's ability to report progress is determined by the external controller's ability to store this in the CR status. If there's a case where progress updates don't happen until the end, then the totalToComplete might just be 1, and the completedSoFar may end up being 0 if not finished, and 1 if finished. I'm not as clear on how the volumesnapshotter handles this. In any case, I'm not sure that's relevant to this statement. It simply says "if the ID is still valid after restart, we should still be able to track progress". If we can't, we can't, of course, and if the ID is no longer valid, then progress calls will error out.

Reply:
> the totalToComplete might just be 1, and the completedSoFar may end up being 0 if not finished, and 1 if finished
My point is this the controller or caller of this func will be confused if the value of `completedSoFar` is the number of snapshots or the number of bytes.  IMO this can be mitigated if we update the struct `OperationProgress` as proposed in[ this comment](https://github.com/vmware-tanzu/velero/pull/5108/files#r1010104191)

Reply:
Resolving this after the fields in `OperationProgress struct` have been updated.